hashing/src/lib.rs

```diff
@@ -113,6 +114,39 @@ impl Digest {
         Digest(result)
     }
 
+    /// Hashes a raw Merkle root and leaf count to firm the final Merkle hash.
+    ///
+    /// To avoid pre-image attacks, the final hash that is based upon the number of leaves in the
+    /// merkle tree and the root hash is prepended with a padding to ensure it is longer than the
+    /// actual chunk size.
+    ///
+    /// Without this feature, an attacker could construct an item that is only a few bytes long but
+    /// hashes to the same value as a much longer, chunked item by hashing `(len || root hash of
+    /// longer items Merkle tree root)`.
+    ///
+    /// This function computes the correct final hash by ensuring the hasher used has been
+    /// initialized with padding before. For efficiency reasons it uses a memoized hasher state
+    /// computed on first run and cloned afterwards.
+    fn hash_merkle_root(leaf_count: u64, root: Digest) -> Digest {
+        static PAIR_PREFIX_HASHER: OnceCell<VarBlake2b> = OnceCell::new();
```

Comment:
Why exactly do we need this? Can't we create an instance of `let mut hasher` right here at the begining of the method?

Reply:
Ah, now I see it:
> For efficiency reasons it uses a memoized hasher state computed on first run and cloned afterwards.

Reply:
I am not so sure. We would be hashing over 8 megabytes (the expected production value for the chunk size) instead of 40 bytes for every single Merkle tree hash we calculate. That is substantial. If our average "large" trie node is 8 megabytes in size, that's a 50% increase on an already CPU heavy operation.

Of course, there's the argument that we won't have _that_ many large trie nodes, but the memoization here took all of 15 minutes to add. So hopefully, this breaks even :)

Reply:
Ah, I totally forgot that we will increase the chunk size to 8MB for TestNet and we will have to put 8MB worth of 0s to protect from preimage attacks.